agent/src/beerocks/slave/tasks/task_pool.h

```diff
@@ -0,0 +1,66 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TASK_POOL_H_
+#define _TASK_POOL_H_
+
+#include "task.h"
+#include <unordered_map>
+
+namespace beerocks {
+
+/**
+ * @brief The TaskPool is single instance tasks container which responsible to run each
+ * task and allow passing messages and events from the Agent to the task.
+ * 
+ */
+class TaskPool {
+public:
+    /**
+     * @brief Add new task to the task pool.
+     * 
+     * @param new_task Shared pointer to the task.
+     * @return true on success, otherwise false.
+     */
+    bool add_task(std::shared_ptr<Task> new_task);
+
+    /**
+     * @brief Run all tasks on the pool, by calling each task work() function.
+     */
+    void run_tasks();
+
+    /**
+     * @brief Send an 'event' defined on a specific task 'task_type'. 
+     * 
+     * @param task_type Task type, defined on Task base class.
+     * @param event Event type, defined on the task itself.
+     */
+    void send_event(Task::eType task_type, uint8_t event);
+
+    /**
+     * @brief Iterate over all tasks on the pool and pass them the message on 'cmdu_rx'.
+     * 
+     * @param cmdu_rx CMDU object containing the message.
+     * @param src_mac MAC address of the message sender.
+     * @param beerocks_header Beerocks header (Only on VS message).
+     * @return true if the message has been handled, otherwise false.
+     */
+    bool handle_cmdu(ieee1905_1::CmduMessageRx &cmdu_rx, sMacAddr src_mac,
+                     std::shared_ptr<beerocks_header> beerocks_header = nullptr);
+
+private:
+    // Decalaring unordered_map with key which is an enum, does not compiles on older gcc version.
+    // It was considered a defect in the standard, and was fixed in C++14, and also fixed in the
+    // version of libstdc++ shipping with gcc as of 6.1.
+    // To make unordered_map work with an enum as key, std::hash<int> function was added as third
+    // template argument.
+    std::unordered_map<Task::eType, std::shared_ptr<Task>, std::hash<int>> m_task_pool;
```

Comment:
No need in Task::eType, see other comments on the issue.
```suggestion
    std::vector<Task*> m_task_pool;
```

Reply:
> No need in Task::eType, see other comments on the issue.

I don't understand your comment. You wrote "No need in Task::eType", but replaced the unordered_map with a vector.
Lookup times are much better in an unordered_map..

Reply:
> how this list is populated?
> I am missing an interface to do so.

https://github.com/prplfoundation/prplMesh/pull/1485/commits/21cc249484f6994b73601a270fd0e3e9331118e7#diff-2adb7ad23cc08452466296d700b98e6dR19

Reply:
My claim is that there is no need to wrap the call to `task.handle_event` with `task_pool.handle_event`.
We gain nothing, and we add more complexity where it is not needed.
The entire enum is redundant. We don't need a map, we don't need to search at all.
We keep the vector for `cmdu` to loop over _all_ tasks.
`send_event`, as declared in this interface just sends the event to the single task that its enum was requested. therefore, the caller already knows who to send the event to, so there is no need to go via the task pool:
It should be as simple as: `task_a.send_event(0x45);`  which is much simpler, faster and cleaner then:
`task_pool.send_event( eTaskA, 0x45);`

Reply:
Task pool is a container of tasks. For example, it allows sending an event to a specific task without having a reference to this task. How do you achieve this without a container?

Reply:
> it allows sending an event to a specific task without having a reference to this task

Q: How?
A: We are keeping a handle (enum) to the task instead of the task itself.

We know at the point of calling the task we want to send a message to.
By sending it to a third party (TaskPool) to call it for us we are doing overwork.
It is much simpler to send the message directly to the task that we want, without the loop via the task-pool. In this way we gain:
* much simpler code - no enum at all.
* expressiveness - we state what we do directly: `task_x.handle(message)`
* run time - no need to search at all.

Reply:
I think that the background is that with the vendor-specific messages, it is possible to include the task ID in the CMDU itself, so that way you can find out which task to send it to. I think... @morantr @vitalybu ?